src/elasticsearch/mongoose_elasticsearch.erl

```diff
@@ -0,0 +1,161 @@
+%%------------------------------------------------------------------
+%% Copyright 2018 Erlang Solutions Ltd.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% @doc Provides functions to interact with ElasticSearch cluster
+%% via HTTP API.
+%%------------------------------------------------------------------
+-module(mongoose_elasticsearch).
+
+-export([start/0]).
+-export([stop/0]).
+-export([health/0]).
+-export([insert_document/4]).
+-export([search/3]).
+-export([count/3]).
+-export([delete_by_query/3]).
+
+-type index() :: binary().
+-type type() :: binary().
+-type document() :: map().
+-type id() :: binary().
+-type query() :: map().
+
+-export_type([index/0]).
+-export_type([type/0]).
+-export_type([document/0]).
+-export_type([id/0]).
+-export_type([query/0]).
+
+-include("mongoose.hrl").
+
+-define(POOL_NAME, elasticsearch).
+
+%%-------------------------------------------------------------------
+%% API
+%%-------------------------------------------------------------------
+
+%% @doc Starts the pool of connections to ElasticSearch cluster.
+%%
+%% Currently only a single pool of connections to a single node may be started. This function pulls
+%% the configuration from `elasticsearch_server' local option. It expects a proplist of following
+%% configuration values:
+%% * `host' (default: `"localhost"') - IP address or a hostname of ElasticSearch node
+%% * `port' (default: `9200') - port number the ElasticSeach node's HTTP endpoint is listening on
+-spec start() -> ignore | ok | no_return().
+start() ->
+    Opts = ejabberd_config:get_local_option(elasticsearch_server),
+    case Opts of
+        undefined ->
+            ignore;
+        _ ->
+            tirerl:start(),
+            start_pool(Opts)
+    end.
+
+%% @doc Stops the pool of connections to ElasticSearch cluster.
+-spec stop() -> ok.
+stop() ->
+    stop_pool(),
```

Comment:
maybe we should have a simple test that start/stop/start works.
Not covered by tests

Reply:
Do you mean to write a big_tests suite which does an RPC and does start -> check health -> stop -> check health -> start -> check health?

Reply:
right

Reply:
I've add `mongoose_elasticsearch_SUITE`. Please check if that's what you meant 🙂